Fix uppercase pinyin (fixing 81)

BorisNA opened the pull request:
Hopefully fix #81 

We can have pinyin with uppercase letters. So we should:
- use IGNORECASE search(), and moreother,
- explicitly name "flags=" argument in "subs", otherwise it is used for the "count" (Uh-oh) (see https://docs.python.org/3/library/re.html#re.sub)
- convert pinyin to lowercase for ruby

BorisNA committed:
Test for uppercase bug #81

- test for main case with uppercase pinyin
- test for ruby (unlike plain pinyin, it should be converted to lowercase)

BorisNA committed:
Fix for uppercase pinyin bug #81

Hopefully fix bug #81

We can have pinyin with uppercase letters. So we should:
- use IGNORECASE search(), and moreother,
- explicitly name "flags=" argument in "subs", otherwise it is used for the "count" (Uh-oh) (see https://docs.python.org/3/library/re.html#re.sub)
- convert pinyin to lowercase for ruby

kieranlblack commented on chinese/color.py:
```diff
@@ -60,14 +60,16 @@ def _repl(p):
             for syllable in word.split():
-                if search(f'^{pattern}$', syllable):
-                    text += sub(f'^{pattern}$', _repl, syllable, IGNORECASE)
+                if search(f'^{pattern}$', syllable, flags=IGNORECASE):
+                    text += sub(f'^{pattern}$', _repl, syllable, flags=IGNORECASE)
```
Nice catch

kieranlblack commented on chinese/color.py:
```diff
@@ -60,14 +60,16 @@ def _repl(p):
                     else:
                         pattern = HALF_RUBY_REGEX
-                    text += sub(pattern, _repl, syllable, IGNORECASE)
+                    # For some reason HALF_RUBY_REGEX is case-insensitive per se, and RUBY_REGEX is case-sensitive
+                    # And for ruby we do not need to keep uppercase
+                    text += sub(pattern, _repl, syllable, flags=IGNORECASE).lower()
```
Hrm, should we make `RUBY_REGEX` also match on `A-Z`? Looking around the code base I don't actually see anywhere that uses this `ruby_whole` branch except for tests... what was supposed to be the use, can we rip it out? I don't use ruby so I don't have the context to know what was originally trying to be done here.

I am a little dubious of the `.lower()` there, like I said I don't use ruby but wouldn't this result in people not being abel to specify pinyin with their own case? I think it someone wanted to not have uppercase they should just not use uppercase rather than have this choice forced on them.

BorisNA replied:
Wow, thanks for reviewing!

1) I tried to look into these regexps, but it's all very confusing because it was edited by different people. Some are inconsistent. In an ideal word all these patterns should be refactored, but... At least for now case sensitivity is handled by the IGNORECASE flag.

2) My rationale for forcing lowercase in ruby is as follows.

We use pinyin like a transcription to get the readings. And as a reading aid (like say hiragana). In the latter case we sometimes have complete sentences in pinyin written western-style with capitalization and so on (for example in lower HSK, children's books). It was exactly my case - I've got an HSK deck with sentences with pre-filled pinyin, and boom, all sentences were marked with the 5th tone.

But for ruby, I believe, it is used strictly as a transcription (like zhuyin/hiragana), never as a complete sentence. So it is only logical to convert it to lowercase to make it consistent and improve readability. I haven't used ruby that much too but decided to give it a try on the back of the cards to speed up reviewing.

> I think it someone wanted to not have uppercase they should just not use uppercase rather than have this choice forced on them.

If one adds pinyin by himself or fills with the addon, yes. But if the pinyin is prefilled from the "good" source, it could already have uppercase. Especially if it is a shared deck.

So, TL;DR - I don't think that somebody will ever want ruby with uppercase letters. But your proposal with adding conversion option to the config looks good and not very difficult.

kieranlblack replied:
Hrm, my preference would be to have no config option at all as it adds extra stuff to maintain / explain to users.

When I look at the Wikipedia page for ruby characters [[0]](https://en.wikipedia.org/wiki/Ruby_character#Examples), their example shows uppercase being used for pinyin which makes me think it might be common to do so. This feels like a personal preference thing outside the scope of this addon that if someone wanted to change then they should do it themselves rather than us trying to do it for them.

I think your patch which fixes the matching of the ruby characters with uppercase pinyin is really great and we should maybe move forward with just that as this will fix colorization for imported decks and such.

[0] https://en.wikipedia.org/wiki/Ruby_character#Examples

BorisNA replied:
Thanks for feedback. Here I'd rather disagree. 
Wikipedia's page is rather strange, I'd say that the example with ruby based on pinyin is incorrect.

Yes, pinyin is capitalized and is written with the word boundaries. But in ruby it is not pinyin for the text/words/sentences (were capitalization makes sense), but rather a kind of hanzi transcription. 

北 per se is not Bei3, but just bei3.

At least in all the books I've seen, there are no capital letters for ruby, if it's really ruby. Sometimes pinyin is written immediately below hanzi (but in this case there will be the whole punctuation, not only syllables, so it differs from ruby).

![image](https://github.com/user-attachments/assets/31f1f5b4-1000-442f-a018-cb8a8dacc32a)

![image](https://github.com/user-attachments/assets/7aaf630f-e2fa-4328-ac24-33fb12049eb6)

kieranlblack commented on tests/test_color.py:
```diff
@@ -35,6 +35,19 @@ def test_separate_syllables(self):
             '<span class="tone4">xiàn</span> <span class="tone4">zài</span>',
         )
 
+    def test_uppercase(self):
```
tests, awesome!

kieranlblack reviewed (commented):
Nice work, want to hear what you say about the choice to force ruby matches into lowercase, but apart from that lgtm.

BorisNA committed:
Config option to force ruby to lowercase

Closes #7

BorisNA committed:
Fix tests for ruby conversion

Testing for config option handling left 'lowercase_ruby" option in inconsistent state.
